net2/HostManager.js

```diff
@@ -1446,7 +1480,43 @@ module.exports = class HostManager {
                   // increment reconnecting count and trigger reconnection
                   updatedPolicy.reconnecting = (updatedPolicy.reconnecting || 0) + 1;
                 }
-                this.setPolicy("vpnClient", updatedPolicy);
+                await this.setPolicyAsync("vpnClient", updatedPolicy);
+                if (fc.isFeatureOn("vpn_disconnect")) {
+                  const broken_time = new Date() / 1000;
+                  setTimeout(async () => {
+                    sem.sendEventToFireApi({
+                      type: 'FW_NOTIFICATION',
+                      titleKey: 'NOTIF_VPN_CLIENT_LINK_BROKEN_TITLE',
+                      bodyKey: 'NOTIF_VPN_CLIENT_LINK_BROKEN_BODY',
+                      titleLocalKey: 'VPN_CLIENT_LINK_BROKEN',
+                      bodyLocalKey: 'VPN_CLIENT_LINK_BROKEN',
+                      bodyLocalArgs: [(settings && (settings.displayName || settings.serverBoxName)) || profileId],
+                      payload: {
+                        profileId: (settings && (settings.displayName || settings.serverBoxName)) || profileId
+                      }
+                    });
+                    const updatedPolicy = this.policy["vpnClient"];
+                    if (!updatedPolicy) return;
+                    if (!updatedPolicy.running) {
+                      const device_cout = await this.getVpnActiveDeviceCount(profileId);
+                      let alarm = new Alarm.VPNDisconnectAlarm(
+                        new Date() / 1000,
+                        null,
+                        {
+                          'p.vpn.profileid': profileId,
+                          'p.vpn.subtype': settings && settings.subtype,
+                          'p.vpn.devicecount': device_cout,
+                          'p.vpn.displayname': (settings && (settings.displayName || settings.serverBoxName)) || profileId,
+                          'p.vpn.time': broken_time,
+                          'p.vpn.strictvpn': settings && settings.strictVPN || false
+                        }
+                      );
+                      await alarmManager2.enqueueAlarm(alarm);
+                      updatedPolicy.waitresume = (settings.strictVPN) ? 1 : 0;
+                      await this.setPolicyAsync("vpnClient", updatedPolicy);
```

Comment:
It's okay to call setPolicy here since it is in the async part of this function.